bundles/org.openhab.binding.magentatv/src/main/java/org/openhab/binding/magentatv/internal/handler/MagentaTVHandler.java

```diff
@@ -0,0 +1,692 @@
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.RefreshType;
+import org.eclipse.smarthome.core.types.State;
+import org.eclipse.smarthome.core.types.UnDefType;
+import org.openhab.binding.magentatv.internal.MagentaTVException;
+import org.openhab.binding.magentatv.internal.MagentaTVGsonDTO.MRPayEvent;
+import org.openhab.binding.magentatv.internal.MagentaTVGsonDTO.MRPayEventInstanceCreator;
+import org.openhab.binding.magentatv.internal.MagentaTVGsonDTO.MRProgramInfoEvent;
+import org.openhab.binding.magentatv.internal.MagentaTVGsonDTO.MRProgramInfoEventInstanceCreator;
+import org.openhab.binding.magentatv.internal.MagentaTVGsonDTO.MRProgramStatus;
+import org.openhab.binding.magentatv.internal.MagentaTVGsonDTO.MRProgramStatusInstanceCreator;
+import org.openhab.binding.magentatv.internal.MagentaTVGsonDTO.MRShortProgramInfo;
+import org.openhab.binding.magentatv.internal.MagentaTVGsonDTO.MRShortProgramInfoInstanceCreator;
+import org.openhab.binding.magentatv.internal.MagentaTVGsonDTO.OAuthAutenhicateResponse;
+import org.openhab.binding.magentatv.internal.MagentaTVGsonDTO.OAuthTokenResponse;
+import org.openhab.binding.magentatv.internal.MagentaTVGsonDTO.OauthCredentials;
+import org.openhab.binding.magentatv.internal.MagentaTVHandlerFactory;
+import org.openhab.binding.magentatv.internal.config.MagentaTVDynamicConfig;
+import org.openhab.binding.magentatv.internal.config.MagentaTVThingConfiguration;
+import org.openhab.binding.magentatv.internal.network.MagentaTVNetwork;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+/**
+ * The {@link MagentaTVHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Markus Michels - Initial contribution
+ */
+@NonNullByDefault
+public class MagentaTVHandler extends BaseThingHandler implements MagentaTVListener {
+    private final Logger logger = LoggerFactory.getLogger(MagentaTVHandler.class);
+    private static final DecimalType ZERO = new DecimalType(0);
+
+    protected MagentaTVDynamicConfig config = new MagentaTVDynamicConfig();
+    private final Gson gson;
+    protected final MagentaTVNetwork network;
+    protected final MagentaTVHandlerFactory handlerFactory;
+    protected MagentaTVControl control = new MagentaTVControl();
+
+    private String thingId = "";
+    private volatile int idRefresh = 0;
+    private @Nullable ScheduledFuture<?> initializeJob;
+    private @Nullable ScheduledFuture<?> pairingWatchdogJob;
+    private @Nullable ScheduledFuture<?> renewEventJob;
+
+    /**
+     * Constructor, save bindingConfig (services as default for thingConfig)
+     *
+     * @param thing
+     * @param bindingConfig
+     */
+    public MagentaTVHandler(MagentaTVHandlerFactory handlerFactory, Thing thing, MagentaTVNetwork network) {
```

Comment:
A handler must not get the handler factory injected - the dependency is only the other way round (the factory keeps track of its handlers).
Please check if you can refactor the code to cleanly decouple the classes. If required, you can think about introducing a "Callback" interface, which is passed in and which could be implemented by the factory, if there's no way around it.

Reply:
I created a new service MagentaTVDeviceManager, review the change and let me know if that is ok

Reply:
Looks ok to me, thanks!